packages/block-editor/src/components/list-view/block-select-button.js

```diff
@@ -55,19 +100,80 @@ function ListViewBlockSelectButton(
 	};
 
 	function onKeyDownHandler( event ) {
-		if ( event.keyCode === ENTER || event.keyCode === SPACE ) {
+		// Handle default mode.
+		if (
+			! labelEditingMode &&
+			( event.keyCode === ENTER || event.keyCode === SPACE )
+		) {
 			onClick( event );
 		}
+
+		// Handle Label editing mode.
+		if ( labelEditingMode ) {
+			// Trap events to input when editing to avoid
+			// default list view key handing (e.g. arrow
+			// keys for navigation).
+			event.stopPropagation();
+
+			// Handle ENTER and ESC exits editing mode.
+			if ( event.keyCode === ENTER || event.keyCode === ESCAPE ) {
+				if ( event.keyCode === ENTER ) {
+					// Submit changes only for ENTER.
+					updateBlockAttributes( clientId, {
+						// Include existing metadata (if present) to avoid overwriting existing.
+						metadata: {
+							...( blockAttributes?.metadata &&
+								blockAttributes?.metadata ),
+							name: inputValue,
+						},
+					} );
+				}
+				toggleLabelEditingMode( false );
+			}
+		}
 	}
 
+	useEffect( () => {
+		if ( labelEditingMode ) {
+			// Focus and select all text on entering label editing mode.
+			inputRef?.current?.focus();
+			inputRef?.current?.select();
+		}
+	}, [ labelEditingMode ] );
```

Comment:
If we figure out how to conditionally show the input field, let's do our best to ensure this effect fires only when the field is showing.

Reply:
👍 

The `labelEditingMode` boolean will determine whether the field is in editable mode. If it's in editable mode then the input will be shown. Note that the focusing/selection does not fire unless `labelEditingMode` is truthy.

Reply:
I've just tested and the focusing of the input only happens if we're in editing mode.